scala-package/core/src/main/scala/org/apache/mxnet/ResourceScope.scala

```
@@ -0,0 +1,195 @@
+  // and https://github.com/scala/scala/blob/2.13.x/src/library/scala/util/Using.scala
+  // TODO: we should move to the Scala util's Using method when we move to Scala 2.13
+  def using[A](scope: ResourceScope = null)(body: => A): A = {
+
+    val curScope = if (scope != null) scope else new ResourceScope()
+
+    val prevScope: Option[ResourceScope] = ResourceScope.getPrevScope()
+
+    @inline def resourceInGeneric(g: scala.collection.Iterable[_]) = {
+      g.foreach( n =>
+        n match {
+          case nRes: NativeResource => {
+            removeAndAddToPrevScope(nRes)
+          }
+          case kv: scala.Tuple2[_, _] => {
+            if (kv._1.isInstanceOf[NativeResource]) removeAndAddToPrevScope(
+              kv._1.asInstanceOf[NativeResource])
+            if (kv._2.isInstanceOf[NativeResource]) removeAndAddToPrevScope(
+              kv._2.asInstanceOf[NativeResource])
+          }
+        }
+      )
+    }
+
+    @inline def removeAndAddToPrevScope(r: NativeResource) = {
+      curScope.remove(r)
+      if (prevScope.isDefined)  {
+        prevScope.get.add(r)
+        r.scope = prevScope.get
+      }
+    }
+
+    @inline def safeAddSuppressed(t: Throwable, suppressed: Throwable): Unit = {
+      if (!t.isInstanceOf[ControlThrowable]) t.addSuppressed(suppressed)
+    }
+
+    var retThrowable: Throwable = null
+
+    try {
+      val ret = body
+       ret match {
+          // don't de-allocate if returning any collection that contains NativeResource.
+        case resInGeneric: scala.collection.Iterable[_] => resourceInGeneric(resInGeneric)
+        case nRes: NativeResource => removeAndAddToPrevScope(nRes)
+        case ndRet: NDArrayFuncReturn => ndRet.arr.foreach( nd => removeAndAddToPrevScope(nd) )
+        case _ => // do nothing
+      }
+      ret
+    } catch {
+      case t: Throwable =>
+        retThrowable = t
+        null.asInstanceOf[A] // we'll throw in finally
+    } finally {
```

Why are we doing all this odd throwable stuff?

---

So, if the code block throws exception we want to still dispose all the NativeResources

---

It was more a question of why aren't we doing something like this instead.
```
} finally {
  try {
    curScope.close
  } catch {
    case closeThrowable: Throwable =>
      if (NonFatal(retThrowable) && !NonFatal(closeThrowable))
        retThrowable = closeThrowable
      else safeAddSuppressed(retThrowable, closeThrowable)
   } finally {
     throw retThrowable
   }
}
```
We call curScope.close regardless of the value of retThrowable so I don't think we need the if. NonFatal(null) returns true so the check to see which to throw should still work. I don't think you need the var toThrow because retThrowable can just be overwritten in the if statement if you decide to throw closeThrowable instead.

I'm on the fence here about whether we should just throw the first exception and not the most fatal exception. My intuition says that the first exception will almost certainly be causing the 2nd regardless of which is fatal.

---

Lets say you get OOM(Native) from code block and then you get a error trying to close all of them, I would still want to throw the OOM exception from the block